crates/store/src/filesystem/mod.rs

```diff
@@ -0,0 +1,412 @@
+// SPDX-FileCopyrightText: 2022 Profian Inc. <opensource@profian.com>
+// SPDX-License-Identifier: AGPL-3.0-only
+
+mod utility;
+
+use super::{Create, CreateError, CreateItem, Error, Get, GetError, Keys, New};
+
+use std::hash::Hash;
+use std::marker::PhantomData;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+use cap_async_std::path::{Path, PathBuf};
+use drawbridge_type::digest::{Algorithms, ContentDigest};
+use drawbridge_type::Meta;
+
+use async_trait::async_trait;
+use cap_async_std::fs::{Dir, File};
+use futures::future::join_all;
+use futures::io::{self, copy, Cursor};
+use futures::stream::{iter, Iter};
+use futures::{AsyncReadExt, AsyncWrite, AsyncWriteExt};
+use mime::Mime;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+
+const KEY_FILE: &str = "key.bson";
+const META_FILE: &str = "meta.bson";
+const CONTENT_LINK_FILE: &str = "content.bin";
+
+/// Filesystem backed storage.
+///
+/// Layout:
+///
+///```text
+///root/
+///  storename/
+///    len-sha256/
+///      key/
+///        (key.bson length)-(key.bson checksum)/
+///          key.bson - unique
+///          meta.bson - possible duplicate
+///          content.bin - symlink to value/*.bin file
+///      value/
+///        (length)-(checksum).bin - unique
+///```
+#[derive(Clone)]
+pub struct Filesystem<K> {
+    root: Dir,
+    key_root: String,
+    value_root: String,
+    phantom: PhantomData<K>,
+}
+
+impl<K> Filesystem<K> {
+    async fn get_keys(&self) -> Vec<Result<K, Error>>
+    where
+        K: DeserializeOwned + 'static + Sync + Send + Clone,
+    {
+        let mut keys = vec![];
+
+        if let Ok(entries) = self.root.read_dir(&self.key_root).await {
+            for key_dir in entries {
+                if let Ok(Ok(key_dir)) = key_dir.map(|dir| dir.open_dir()) {
+                    if let Ok(entries) = key_dir.entries().await {
+                        for key_file in entries {
+                            let key_file = match key_file {
+                                Err(_) => continue,
+                                Ok(key_file) => key_file,
+                            };
+
+                            if utility::dir_entry_contains(&key_file, "key.bson") {
+                                let key_file = key_file.open();
+
+                                keys.push(async {
+                                    self.read_bincode_file(&mut key_file.map_err(Error::Io)?)
+                                        .await
+                                });
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        join_all(keys.into_iter()).await
+    }
+
+    async fn get_key_dir(&self, expected_identify: &str) -> Result<PathBuf, Error>
```

Comment:
Is `identify` here a typo for `identity`?

Reply:
yes